### ISSUE TITLE
When is 'assert' used in why3?

### ISSUE THREAD

**Issue by Re-st** [kaist-cp/cs220#138]
Looking at [why3 docs](https://why3.lri.fr/manual.pdf), I don't get a clear understanding of "assert"s are..

Comparing line 1-39 (has assert) vs 44-76 (use no assert, uses requires-ensures pair instead) in https://gitlab.inria.fr/why3/why3/-/blob/master/examples/binary_search.mlw, assert kinda 'feels' like some middle-point checking, but I can't get a clear grasp...

And since in some codes 'asserts' were necessary to go green, it doesn't seem like a debug-help function either. It gives more confusion.

When do we use asserts, and what are the advantages?

**Comment by minseongg** [comment]
From the page 146 of [why3 documentation](https://why3.lri.fr/manual.pdf):

`assert` creates an intermediate subgoal. This is comparable to assert written in WhyML code. Here, the intent is only to help provers by specifying one key argument of the reasoning they should use.

Example: From a goal of the form $\Gamma \vdash G$, the transformation assert (n = 0) produces the following two tasks: $\Gamma \vdash h: n = 0$ and $\Gamma, h: n = 0 \vdash G$. This effectively adds h as an intermediate goal to prove.

**Comment by 7dy109** [comment]
In my case, I used assert when I want to know what why3 can verify the things at that time when the assert is introduced. This helped me to solve assignment 5.
And also, I used assert when I want to verify program which invariant cannot.
For example, in that binary search code, if assert { l <= m <= u }; change to invariant { l <= m <= u };, why3 doesn't verify the program because at the end of the loop it is possible that 'l > u'
But I'm not sure whether this is the advantage of assert.
